### ISSUE TITLE
Type narrowing should account for nullish coalescing operator in conditions.

### ISSUE THREAD

**Issue by oleg-slapdash** [microsoft/TypeScript#45941]
# Bug Report

### 🔎 Search Terms

type-narrowing, nullish coalescing

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I did not review FAQ since I don't know which section to review.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=4.5.0-dev.20210910#code/MYewdgzgLgBAbgQwE4EsECMA2BTGBeGAWQSgAsA6JBMAExAFsAKAShgD4YAGcgVhgH4YAIjAhYAV1rYAZimy0hMAFwxJNGSjDYaAbgCwAKEMppMRolQYcAwWo1aarAN6GYbmAHoPMAKJIkIEgqAPLoAFbYwLAoEDAADiAQEChYAJ4wAOR2sg4Z5IwATDwAzAXMru4WaFjY5DhgAOZkhgC+hsamjACE5sjV1vyCImKqUrLyNELMzjAtMNiYELguBu6e3n4BQTChEVEwMfGJyWmZ2ZraeYUlZRVuVVa19U2kre0GoJCwYqTYSABqfUe+HgQJqNmEogkYzkCg6Zi6Pz+gMsNRmcwWSxgKzWXhgAHVAgBrCB3GBIgFgnB1eQvN5GAxAA

### 💻 Code

```typescript
const variable = Math.random() > 0.5 ? "not undefiend" : undefined;

if (variable ?? undefined) {
    // Error: Object is possibly 'undefined'.(2532)
    variable.length
}

if (!(variable ?? "not undefiend")) { } else {
    // Error: Object is possibly 'undefined'.(2532)
    variable.length
}

const otherVariable = variable ?? "not undefiend"
if (!otherVariable) { } else {
    // Works
    otherVariable.length
}


```

### 🙁 Actual behavior

Type narrowing fails to detect that instance cannot be undefined.

### 🙂 Expected behavior

TS should detect that nullish coalescing will always be true or false if the variable is undefined and hence cannot be undefined.

**Comment by lubieowoce** [comment]
This snippet (example #1) is the only one that shows the (minor) problem:

```
declare const x: object | undefined
if (x ?? undefined) {
   // x should be narrowed to object here, but isn't
}
```
But this doesn't really matter -- the `?? undefined` doesn't change the code's behavior in any way and should be removed anyway.

Examples 2 and 3 don't show the problem. You can see that when you simplify them a bit by removing the double negations (i.e. remove `!`, swap branches):

```typescript
if (variable ?? "not undefiend") {
    // Should fail, fails (because we'd still get here if `variable === undefined`)
    variable.length
}

const otherVariable = variable ?? "not undefiend"
if (otherVariable) {
    // Should work, works (because otherVariable: string)
    otherVariable.length
}
```

i played around with it a bit and it seems typescript doesn't like narrowing types on `if`s like these in general: [see playground](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210910#code/CYUwxgNghgTiAEYD2A7AzgF3gDwFzyQCMArcLAH3gFcVQAzASxRGAFgAoDgei-gAsQEAA4gYaeBiTw0fJAHcJATxHwUsGPKYBzeKI1iOoSLATJ0WIsXwAKPARJkAlPAC8APngA3JA2DxD4NBwiKiY1LQgdDZ2NPRMLM7uXj5+HNy8AGIAqgAyOQCa8ADCAPIASmUAokUAKmnsDHTwtvDklHRQEGggzgDeHPCD9sS2zjyq6vIsElKWZACEHAC+ul0I-exD4fSj8ONqMBpy05LbkfHAi+xL9fXjAAoAgmU1AJKPeYWlFdV1nA1NFoAMiB8AAjH0BkNLLt9pNjn5TnMwBgritBN14BstrFIrDeDIkFQIH5CAgDkcTlJcYxmMAADTwQhULAMNAoADkGGW9UazWw8BBWKWkM20JI+ImhymiNmDhRaNWmOxQxpksJxNJ5PhVLOtJYjOZrPZXJ5-zuvFeADlvlVarzAQKAPxOvUXUVbGHYMYE2Sapna6UImbDMiGlnwNmc7nXJXrKGDNXeva8CkykM0i5oh38+Aut10j3ikbJ8YakkBqWU2WhlHh43R5ZxrEJvWStPB06ZunZ-58lptAsJFtiwZen3SP0VslV9NI+UYeuRk0x9FrEc4iJ0ds6mvdli9jj9gWDlDEiBFscS0u+onTwPVkPIxdMiNR02xjHx0dtm+zzvUluWZmkejqtJQAAMl7DOqU5av+urPku76rs2KqJluO5Brq+6XCB-xAA). the only case where it worked is:

```
if (x || false) {
    x.foo // narrowed to object
} else {
    x // narrowed to undefined
}
```

**Comment by oleg-slapdash** [comment]
@lubieowoce It seems like you are suggesting that this should be the linter`s job to detect this. I guess my bug report can be narrowed down to this snippet:
```typescript
declare const var1: string[] | undefined;
if (var1 || false) {
    // Works
    var1.length
}
if (var1 ?? false) {
    // Object is possibly 'undefined'.
    var1.length
}
```

Typescript already detects the first case and should detect the second case as well.

**Comment by andrewbranch** [comment]
Why would you write `|| false` _or_ `?? false`? Does this ever manifest in realistic code?

**Comment by lubieowoce** [comment]
@andrewbranch yeah, that's what i meant by "should be removed anyway" in my original comment. but i guess this could come up in some library scenario, e.g. some flexible options API, and you want to collapse many "absent" states into one:
```
declare const wide: object | undefined | null | false;
const narrow = wide || null // object | null
```

**Comment by andrewbranch** [comment]
Closing in absence of a convincing explanation for why this would matter in real life. Let me know if I’ve missed something.
